code/datums/emote.dm

```diff
@@ -0,0 +1,408 @@
+	if(!can_run_emote(user, TRUE, intentional))
+		return FALSE
+	var/msg = select_message_type(user, message, intentional)
+	if(params && message_param)
+		msg = select_param(user, params)
+
+	msg = replace_pronoun(user, msg)
+
+	if(!msg)
+		return
+
+	if(isobserver(user))
+		log_ghostemote(msg, user)
+	else
+		log_emote(msg, user)
+
+	var/dchatmsg = "<b>[user]</b> [msg]"
+
+	var/tmp_sound = get_sound(user)
+	// If our sound emote is forced by code, don't worry about cooldowns at all.
+	if(tmp_sound && should_play_sound(user, intentional) && (!intentional || !user.start_audio_emote_cooldown(type, audio_cooldown)))
+		if(age_based && istype(user, /mob/living/carbon/human))
+			var/mob/living/carbon/human/H = user
+
+			playsound(user, tmp_sound, volume, vary, frequency = H.get_age_pitch())
+		else
+			playsound(user, tmp_sound, volume, vary)
+
+	var/user_turf = get_turf(user)
+	if (user.client)
+		for(var/mob/ghost as anything in GLOB.dead_mob_list)
+			if(!ghost.client || isnewplayer(ghost))
+				continue
+			if(ghost.client.prefs.toggles & PREFTOGGLE_CHAT_GHOSTSIGHT && !(ghost in viewers(user_turf, null)))
+				ghost.show_message("<span class='emote'>[ghost_follow_link(user, ghost)] [dchatmsg]</span>")
+
+	if(emote_type & EMOTE_VISIBLE || user.mind?.miming)
+		user.audible_message(dchatmsg, deaf_message = "<span class='emote'>You see how <b>[user]</b> [msg]</span>")
+	else
+		user.visible_message(dchatmsg, blind_message = "<span class='emote'>You hear how <b>[user]</b> [msg]</span>")
+
+	if(!(emote_type & (EMOTE_FORCE_NO_RUNECHAT | EMOTE_SOUND)))
+		var/runechat_text = msg
+		if(length(msg) > 100)
+			runechat_text = "[copytext(msg, 1, 101)]..."
+		var/list/can_see = get_mobs_in_view(1, user)  //Allows silicon & mmi mobs carried around to see the emotes of the person carrying them around.
+		can_see |= viewers(user,null)
+		for(var/mob/O in can_see)
+			if(O.status_flags & PASSEMOTES)
+				for(var/obj/item/holder/H in O.contents)
+					H.show_message(message, EMOTE_VISIBLE)
+
+				for(var/mob/living/M in O.contents)
+					M.show_message(message, EMOTE_VISIBLE)
+
+			if(O.client?.prefs.toggles2 & PREFTOGGLE_2_RUNECHAT)
+				O.create_chat_message(user, runechat_text, symbol = RUNECHAT_SYMBOL_EMOTE)
+
+
+	SEND_SIGNAL(user, COMSIG_MOB_EMOTED(key), src, key, emote_type, message, intentional)
+	SEND_SIGNAL(user, COMSIG_MOB_EMOTE, key, intentional)
+
+/**
+ * For handling emote cooldown, return true to allow the emote to happen.
+ *
+ * Arguments:
+ * * user - Person that is trying to send the emote.
+ * * intentional - Bool that says whether the emote was forced (FALSE) or not (TRUE).
+ *
+ * Returns FALSE if the cooldown is not over, TRUE if the cooldown is over.
+ */
+/datum/emote/proc/check_cooldown(mob/user, intentional)
+	if(!intentional)
+		return TRUE
+	if(user.emotes_used && user.emotes_used[src] + cooldown > world.time)
+		return FALSE
+	if(!user.emotes_used)
+		user.emotes_used = list()
+	user.emotes_used[src] = world.time
+	return TRUE
```

Comment:
this proc needs to return FALSE if `EMOTE_ON_COOLDOWN`

Reply:
I think this is actually a separate cooldown, the emote cooldown defines relate to audio. I might want to clarify that

Reply:
Fixed in https://github.com/ParadiseSS13/Paradise/pull/17544/commits/05b2c74ca8a36f49521c869c180722f2c63b48ff